File: starter-code/python_example_telemetry_api_extension/python_example_telemetry_api_extension/telemetry_dispatcher.py
import os
import requests
import json
from queue import Queue

AWS_LAMBDA_FUNCTION_NAME = os.environ.get('AWS_LAMBDA_FUNCTION_NAME')

#Environment variables used to configure extension function
DISPATCH_POST_URI = os.getenv("DISPATCH_POST_URI")
ENDPOINT_USER = os.getenv("ENDPOINT_USER")
ENDPOINT_PASSWORD = os.getenv("ENDPOINT_PASSWORD")

def dispatch_telemetry(queue, force):
    while (not queue.empty()):
        print ("[telemetry_dispatcher] Dispatch telemetry data")
        filtered_queue_metrics = []
        filtered_queue_logs = []
        batch = queue.get_nowait()

       #iterate through events in the batch
        for event in batch:
            #get event types that we are interested in -- 1/ function logs 2/ execution reports 
            if event.get('type') == "function":
                filtered_queue_logs.append(event)
            elif event.get('type') == "platform.report":
                filtered_queue_metrics.append(event)

        #Transform data to match backend API
        log_data = format_log_data(filtered_queue_logs)
        metric_data = format_metric_data(filtered_queue_metrics)

        #Dispatch data to backend API 
        post_data(log_data,'log')
        post_data(metric_data,'metric')

def post_data(data,endpoint):

    #OUR CODE GOES HERE

    return True

def format_log_data(filtered_queue):

    #OUR CODE GOES HERE

    return True

def format_metric_data(filtered_queue):

    #OUR CODE GOES HERE

    return True

File: starter-code/nodejs/nodejs_example_telemetry_api_extension/telemetry_dispatcher.js
const fetch = require('node-fetch');

const AWS_LAMBDA_FUNCTION_NAME = process.env.AWS_LAMBDA_FUNCTION_NAME;
const DISPATCH_POST_URI = process.env.DISPATCH_POST_URI;
const ENDPOINT_USER = process.env.ENDPOINT_USER;
const ENDPOINT_PASSWORD = process.env.ENDPOINT_PASSWORD;

function dispatchTelemetry(queue, force) {
    while (queue.length !== 0) {
        console.log('[telemetry_dispatcher] Dispatch telemetry data');
        const filteredQueueMetrics = [];
        const filteredQueueLogs = [];
        const event = queue.shift();

        if (event.type === 'function') {
            filteredQueueLogs.push(event);
        } else if (event.type === 'platform.report') {
            filteredQueueMetrics.push(event);
        }

        const logData = formatLogData(filteredQueueLogs);
        const metricData = formatMetricData(filteredQueueMetrics);
        postData(logData, 'log');
        postData(metricData, 'metric');
    }
}

function formatLogData(filteredQueue) {
    //OUR CODE GOES HERE
}

function formatMetricData(filteredQueue) {
    //OUR CODE GOES HERE
}

async function postData(data, endpoint) {
    //OUR CODE GOES HERE
}

module.exports = {
    dispatchTelemetry
};